zookeeper/src/main/java/com/linecorp/armeria/client/endpoint/zookeeper/ZookeeperEndpointGroup.java

```diff
@@ -0,0 +1,256 @@
+                    } catch (Exception e) {
+                        throw new EndpointGroupException("Failed to process ZooKeeper watch event", e);
+                    }
+                }
+            }
+        }
+
+        @Override
+        public void processResult(int responseCodeInt, String path, Object ctx, Stat stat) {
+            boolean exists;
+            Code responseCode = get(responseCodeInt);
+            switch (responseCode) {
+                case OK:
+                    exists = true;
+                    break;
+                case NONODE:
+                    exists = false;
+                    break;
+                case SESSIONEXPIRED:
+                    //ignore this and let the zNode Watcher process it first
+                case NOAUTH:
+                    //possible happens during application runtime, we ignore this and wait
+                    //for the ACL returned to normal. When it happens at application Start time, the
+                    //ZookeeperEndpointGroup class constructor will throw an exception
+                    return;
+                default:
+                    //recoverable errors, retry it, fatal errors goes to Watcher process routine
+                    try {
+                        zkHandleProxy.get().exists(zNode, true, this, null);
+                    } catch (Exception ex) {
+                        throw new EndpointGroupException("Failed to process ZooKeeper callback event", ex);
+                    }
+                    return;
+            }
+            byte[] nodeData = null;
+            if (exists) {
+                try {
+                    nodeData = zkHandleProxy.get().getData(zNode, false, null);
+                } catch (Exception e) {
+                    logger.warn("error to get zNode data:  " + e.getMessage());
+                    //just return, fatal errors goes to Watcher process routine
+                    return;
+                }
+            }
+            if (nodeData == null && nodeData != prevData ||
+                nodeData != null && !Arrays.equals(prevData, nodeData)) {
+                prevData = nodeData;
+                try {
+                    endpointList = converter.convert(prevData);
+                } catch (IllegalArgumentException exception) {
+                    logger.warn("Failed to convert zNode value to EndpointGroup:  " + exception.getMessage() +
+                                ", invalid value:  " + new String(prevData, Charsets.UTF_8));
+                }
+            }
+        }
+
+        private void recoverZkConnection() {
+            //recoverZkConnection  by using a exponential backoff strategy
+            try {
+                Thread.sleep(retryDelay);
+            } catch (InterruptedException e) {
+                throw new EndpointGroupException("Failed to recover ZooKeeper connection", e);
+            }
+            retryDelay = Math.min(MAX_RETRY_DELAY, retryDelay * 2);
```

Comment:
You should not increase the retry delay yet. You should only when `doConnect()` fails.